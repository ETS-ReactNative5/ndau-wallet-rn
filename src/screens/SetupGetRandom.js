import React, { Component } from 'react';
import { StyleSheet, View, Text, ScrollView } from 'react-native';
import Base64 from 'base-64';
import CommonButton from '../components/CommonButton.js';
import Randal from '../helpers/randal.js';
import Stepper from '../components/Stepper';
import cssStyles from '../css/styles';
import SetupStore from '../model/SetupStore';
import { SafeAreaView } from 'react-navigation';

class SetupGetRandom extends Component {
  constructor(props) {
    super(props);
    this.state = {
      entropy: 'ZWEQAwQFBgcICQoLDA0ODw==',
      percentage: 0,
      doneDisabled: true,
      scribbling: false
    };

    this.randalPromise = this.initRandal().catch((e) => {
      console.error(e);
    });
  }

  initRandal() {
    this.randal = new Randal();
    return this.randal.init().then(() => {
      this.randal.onUpdate(() => {
        this.setState({
          entropy: Base64.encode(this.randal.getHash().substr(0, 16)),
          percentage: this.randal.getPercentage()
        });
      });

      this.randal.onDone(() => {
        this.setState({ doneDisabled: false });
      });
    });
  }

  showNextSetup = () => {
    SetupStore.setEntropy(this.state.entropy);
    this.props.navigation.navigate('SetupYourWallet');
  };

  handleScribble(evt) {
    this.onScribbleStart();
    this.randal.checkPoint(evt.nativeEvent.locationX, evt.nativeEvent.locationY);
  }
  onScribbleStart() {
    if (this.state.scribbling) {
      return;
    }
    this.setState({ scribbling: true });
  }
  onScribbleEnd() {
    if (!this.state.scribbling) {
      return;
    }
    this.setState({ scribbling: false });
  }

  onScribbleStart() {
    if (this.state.scribbling == true) {
      return;
    }
    this.setState({ scribbling: true });
  }
  onScribbleEnd() {
    if (this.state.scribbling == false) {
      return;
    }
    this.setState({ scribbling: false });
  }

  render() {
    const { scribbling } = this.state;
    return (
      <SafeAreaView style={styles.safeContainer}>
        <View style={cssStyles.container}>
          <ScrollView style={styles.contentContainer} scrollEnabled={!scribbling}>
            <Stepper screenNumber={3} />

            <View>
              <Text style={cssStyles.wizardText}>
                To generate the strongest possible encryption, we need a source of random input.
                Scribble in the box below to add randomness to your key.
              </Text>
            </View>
            <View style={{ marginTop: 8 }}>
              <ProgBar percentage={this.state.percentage} />
              <View
                onStartShouldSetResponderCapture={() => true}
                onResponderMove={(evt) => this.handleScribble(evt)}
                style={styles.scribbleArea}
                onResponderRelease={() => this.onScribbleEnd()}
                onResponderTerminate={() => this.onScribbleEnd()}
              />
            </View>
          </ScrollView>
          <View style={styles.footer}>
            <CommonButton
              onPress={this.showNextSetup}
              title="Done"
              disabled={this.state.doneDisabled}
            />
          </View>
        </View>
      </SafeAreaView>
    );
  }
}

function ProgBar(props) {
  let percentage = Math.min(props.percentage, 100);
  return (
    <View
      style={{
        height: 20,
        flex: 1,
        backgroundColor: 'grey',
        marginBottom: 20,
        marginTop: 10
      }}
    >
      <View
        style={{
          height: 20,
          backgroundColor: percentage == 100 ? '#4e957a' : '#f99d1c',
          width: String(percentage) + '%'
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  safeContainer: {
    flex: 1,
    backgroundColor: '#1c2227'
  },
  contentContainer: {
    flex: 1 // pushes the footer to the end of the screen
  },
  footer: {
    justifyContent: 'flex-end'
  },
  progress: {
    paddingTop: 30,
    paddingBottom: 30
  },
  scribbleArea: {
    backgroundColor: 'white',
    flex: 1,
    height: 200
  }
});

export default SetupGetRandom;
