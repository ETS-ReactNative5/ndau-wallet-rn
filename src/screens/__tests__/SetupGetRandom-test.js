jest.mock('react-native-securerandom');

import { mount } from 'enzyme';
import React from 'react';
import { StyleSheet } from 'react-native';
import SetupGetRandom from '../SetupGetRandom';
import renderer from 'react-test-renderer';

const makeStyles = () =>
  StyleSheet.create({
    wizardText: {
      color: '#ffffff',
      fontSize: 20
    }
  });
const navigator = {
  setStyle: () => {},
  toggleNavBar: () => {}
};

describe('SetupGetRandom snapshot', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<SetupGetRandom navigator={navigator} parentStyles={makeStyles()} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('SetupGetRandom behavior', () => {
  beforeEach((done) => {
    // set up default wrapper for every test.
    this.wrapper = mount(<SetupGetRandom navigator={navigator} parentStyles={makeStyles()} />);
    this.wrapper.instance().randalPromise.then(done);
  });

  it('done button starts disabled', () => {
    const onlyButton = this.wrapper.find('Button').at(0);
    expect(onlyButton.prop('disabled')).toBeTruthy();
  });

  it('done button starts disabled', () => {
    const onlyButton = this.wrapper.find('Button').at(0);
    expect(onlyButton.prop('disabled')).toBeTruthy();
  });

  // This calls a handler method on the component with mock data and the scope.
  // Then it updates the wrapper and inspects the subcomponent's properties.
  it('after scribbling done button is enabled', () => {
    const instance = this.wrapper.instance();
    const handler = instance.handleScribble;
    let flipper = -1;
    const x = 500; // random position
    // scribble for an arbitrary amount of time
    for (var i = 0; i < 2048; i++) {
      handler.apply(instance, [
        {
          nativeEvent: {
            locationX: x * flipper,
            locationY: x * flipper
          }
        }
      ]);
      flipper *= -1;
    }
    this.wrapper.update(); // important
    const doneButton = this.wrapper.find('Button').at(0);
    expect(doneButton.prop('disabled')).toBeFalsy();
  });

  it('subsequent scribbles should not produce the same output', (done) => {
    const tickleScribbler = (wrapper) => {
      const instance = wrapper.instance();
      const handler = instance.handleScribble;

      let flipper = -1;
      const x = 500; // random position
      // scribble for an arbitrary amount of time
      for (var i = 0; i < 2048; i++) {
        handler.apply(instance, [
          {
            nativeEvent: {
              locationX: x * flipper,
              locationY: x * flipper
            }
          }
        ]);
        flipper *= -1;
      }
    };

    anotherWrapper = mount(<SetupGetRandom parentStyles={makeStyles()} />);
    anotherWrapper.instance().randalPromise.then(() => {
      tickleScribbler(this.wrapper);
      tickleScribbler(anotherWrapper);
      try {
        const hash = this.wrapper.instance().randal.getHash();
        const anotherHash = anotherWrapper.instance().randal.getHash();
        expect(hash).not.toBe(anotherHash);
        done();
      } catch (e) {
        done.fail(e);
      }
    });
  });
});
