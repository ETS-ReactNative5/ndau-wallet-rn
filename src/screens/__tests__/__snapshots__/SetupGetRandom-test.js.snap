// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing SetupGetRandom... renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Object {
      "backgroundColor": "#1c2227",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#1c2227",
        "flex": 1,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 20,
              }
            }
          >
            Get random
          </Text>
        </View>
        <View>
          <RCTProgressView
            progress={0.375}
            style={
              Array [
                Object {
                  "height": 2,
                },
                undefined,
              ]
            }
          />
        </View>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#ffffff",
                "fontSize": 20,
              }
            }
          >
            To generate the strongest possible encryption, we need a source of random input. Scribble in the box below to add randomness to your key.
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          placeholder="Scribble area"
          placeholderTextColor="#333"
          style={
            Object {
              "borderColor": "gray",
              "borderWidth": 1,
              "color": "#ffffff",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 20,
              "height": 200,
              "marginBottom": 10,
              "marginTop": 10,
              "paddingLeft": 10,
            }
          }
          value="133755ff123456789111222333444555"
        />
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessibilityComponentType="button"
        accessibilityTraits={
          Array [
            "button",
          ]
        }
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "padding": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#4d9678",
                },
              ]
            }
          >
            Done
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
